### ISSUE TITLE
[REVIEW]: EvoMaster: A Search-Based System Test Generation Tool

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2153]
**Submitting author:** @arcuri82 (<a href="http://orcid.org/0000-0003-0799-2930">Andrea Arcuri</a>)
**Repository:** <a href="https://github.com/EMResearch/EvoMaster" target ="_blank">https://github.com/EMResearch/EvoMaster</a>
**Version:** 1.1.0
**Editor:** @gkthiruvathukal
**Reviewers:** @mado89, @s0nata, @UTH-Tuan
**Archive:** <a href="https://doi.org/10.5281/zenodo.4300745" target="_blank">10.5281/zenodo.4300745</a>

## Status

[![status](https://joss.theoj.org/papers/c58344ed30444a648b60b5e534fa1fed/status.svg)](https://joss.theoj.org/papers/c58344ed30444a648b60b5e534fa1fed)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/c58344ed30444a648b60b5e534fa1fed"><img src="https://joss.theoj.org/papers/c58344ed30444a648b60b5e534fa1fed/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/c58344ed30444a648b60b5e534fa1fed/status.svg)](https://joss.theoj.org/papers/c58344ed30444a648b60b5e534fa1fed)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/EMResearch/EvoMaster" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@mado89, @UTH-Tuan, @s0nata, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @gkthiruvathukal know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/EMResearch/EvoMaster">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@arcuri82) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/EMResearch/EvoMaster">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@arcuri82) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/EMResearch/EvoMaster">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@arcuri82) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @mado89, @jaynanavati it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- None

MISSING DOIs

- https://doi.org/10.1109/icst.2018.00046 may be missing for title: EvoMaster: Evolutionary Multi-context Automated System Test Generation
- https://doi.org/10.1007/978-3-319-66299-2_1 may be missing for title: Many Independent Objective (MIO) Algorithm for Test Suite Generation
- https://doi.org/10.1109/qrs.2017.11 may be missing for title: RESTful API Automated Test Case Generation
- https://doi.org/10.1145/3321707.3321815 may be missing for title: Resource-based test case generation for RESTful web services
- https://doi.org/10.1145/3321707.3321732 may be missing for title: SQL data generation to enhance Search-Based System Testing

INVALID DOIs

- None
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02153/joss.02153/10.21105.joss.02153.pdf)

**Comment by arfon** [comment]
Dear authors and reviewers

We wanted to notify you that in light of the current COVID-19 pandemic, JOSS has decided to suspend submission of new manuscripts and to handle existing manuscripts (such as this one) on a "best efforts basis". We understand that you may need to attend to more pressing issues than completing a review or updating a repository in response to a review. If this is the case, a quick note indicating that you need to put a "pause" on your involvement with a review would be appreciated but is not required.

Thanks in advance for your understanding.

_Arfon Smith, Editor in Chief, on behalf of the JOSS editorial team._

**Comment by mado89** [comment]
@arfon Sorry to hear about the closing down. Hopefully, once my submission is ready everything is back to normal
@arcuri82 just wanted to let you know I haven't forgotten about this. I am a bit behind schedule but I hope to start on the weekend or next week. Sorry for the delay

**Comment by arfon** [comment]
:wave: @mado89, @jaynanavati - just a friendly check-in to see how things are going with your reviews?

**Comment by mado89** [comment]
I see that @arcuri82 is working on this repository. Which version should be reviewed? 1.0.0 or the snapshot for 1.0.2?

**Comment by arfon** [comment]
@arcuri82 - could you advise if there are any major changes between `v1.0.0` and `v1.0.2`?

**Comment by arcuri82** [comment]
the project is an active development (as it has been for the last 3.5 years).
The latest release is `v1.0.1` (see [release page](https://github.com/EMResearch/EvoMaster/releases)), and it was made 2 weeks ago. It is just bug fixes over `v1.0.0`.

The version `1.0.2-SNAPSHOT` is simply the working version in the master branch.

Which one to review? Both would be equivalent, as all new features under development are deactivated by default. It all depends if you want to download an already packaged jar file from the release page, or build the jar directly

**Comment by arcuri82** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02153/joss.02153/10.21105.joss.02153.pdf)

**Comment by arcuri82** [comment]
thanks @mado89 for your feedback. i made changes and updated the paper. Let me know if you require further changes

Question to @arfon: is @jaynanavati active? I was looking at his/her GitHub profile today, and it does not seem there is any activity whatsoever on GitHub:  https://github.com/jaynanavati?tab=overview&from=2020-04-01&to=2020-04-24
Or did I miss something?  (I apologize in that case)

**Comment by arfon** [comment]
> Question to @arfon: is @jaynanavati active? I was looking at his/her GitHub profile today, and it does not seem there is any activity whatsoever on GitHub: https://github.com/jaynanavati?tab=overview&from=2020-04-01&to=2020-04-24

Not clear. I'd like us to wait longer before considering changing out reviewers, especially as so many people's lives are severely disrupted by COVID-19.

**Comment by arcuri82** [comment]
COVID-19 is indeed a problem :(
However, what is peculiar here is that @jaynanavati account was created in 2018, and there is absolutely no activity of any kind since then (unless all activities are in private repositories, and then those are not shown). 
Anyway, I am not in a hurry :)

**Comment by arcuri82** [comment]
I would like to apologize: I forgot (or did not pay proper attention) of the "[PRE REVIEW]" #2001 , where @jaynanavati explicitly stated to be willing to be a reviewer. Somehow, I was worried that a wrong user was added by mistake to this review, and just wanted to double-check :)

**Comment by jaynanavati** [comment]
Dear Editor,

Kindly replace me with someone else as I don't have spare time to review
the work nowadays.

Thank you.

Regards,

Dr. Jay Nanavati

**Comment by arfon** [comment]
@arcuri82 - looks like we're going to need to find another reviewer - could you take a look a [this list of potential reviewers](https://bit.ly/joss-reviewers) and identify a few people who would be good candidates to review this submission?

**Comment by arcuri82** [comment]
@arfon - considering the ones that have been already contacted in #2001, then some other possibilities could be:
- @feliperodri
- @hainesr
- @s0nata
- @UTH-Tuan
- @shanamatthews
thanks

**Comment by hainesr** [comment]
Hi all,

I am interested in this work, but I am terribly sorry I can't see that I could review it in a timely manner at the moment. Home schooling is currently a large drain on my resources!

Rob

**Comment by mado89** [comment]
> thanks @mado89 for your feedback. i made changes and updated the paper. Let me know if you require further changes
I will take a look at it. Sorry once again for being so slow on this review. Hopefully, the second reviewer - once found - can do their job faster than me so that you won't be delayed for much longer

**Comment by arfon** [comment]
@whedon remove @jaynanavati as reviewer

**Comment by whedon** [comment]
OK, @jaynanavati is no longer a reviewer

**Comment by arfon** [comment]
👋 @feliperodri & @s0nata - we're looking for a second reviewer of the [EvoMaster](https://github.com/EMResearch/EvoMaster) package for JOSS.

Would either of you be willing to review this submission for us? We carry out our checklist-driven reviews here in GitHub issues and follow these guidelines: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html

**Comment by feliperodri** [comment]
Hi @arfon, thanks for the invitation. I can take some time this week to review it and provide a feedback by Friday or Saturday. Would that be a reasonable time frame for you?

**Comment by arfon** [comment]
> Hi @arfon, thanks for the invitation. I can take some time this week to review it and provide a feedback by Friday or Saturday. Would that be a reasonable time frame for you?

That would be fantastic, thanks! I'll go ahead and set up your review here now.

**Comment by arfon** [comment]
@whedon add @feliperodri as reviewer

**Comment by whedon** [comment]
OK, @feliperodri is now a reviewer

**Comment by arfon** [comment]
@feliperodri - please carry out your review in this issue by updating the checklist above and giving feedback in this issue. The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html

Any questions/concerns please let me know.

**Comment by s0nata** [comment]
hello @arfon, thanks for the invitation, I should have some time over the weekend.

**Comment by arfon** [comment]
> hello @arfon, thanks for the invitation, I should have some time over the weekend.

OK great! I'll set up your review now...

**Comment by arfon** [comment]
@whedon add @s0nata as reviewer

**Comment by whedon** [comment]
OK, @s0nata is now a reviewer

**Comment by arfon** [comment]
@s0nata - please carry out your review in this issue by updating the checklist above and giving feedback in this issue. The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html

Any questions/concerns please let me know.

**Comment by arfon** [comment]
@gkthiruvathukal - feel free to take this over when you are ready.

**Comment by s0nata** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02153/joss.02153/10.21105.joss.02153.pdf)

**Comment by s0nata** [comment]
Hi @arcuri82, I have just added my comments on the paper text to the issue [#170](https://github.com/EMResearch/EvoMaster/issues/170)

**Comment by arcuri82** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02153/joss.02153/10.21105.joss.02153.pdf)

**Comment by arcuri82** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02153/joss.02153/10.21105.joss.02153.pdf)

**Comment by gkthiruvathukal** [comment]
@arfon Just letting you know, I'm able to take over and see this through.

@mado89, @feliperodri, @s0nata, How are things coming along with the review? Let me know if you have any questions.

**Comment by s0nata** [comment]
@gkthiruvathukal almost ready on my end, just need to fix an issue with my OS that blocks tool evaluation, and I'm done

**Comment by gkthiruvathukal** [comment]
Thanks, @s0nata!  Please signal me when ready!

**Comment by arcuri82** [comment]
Sadly, today marks 6 months since submission (7th January)... :(

**Comment by gkthiruvathukal** [comment]
@arcuri82 So sorry about this. I am still hopeful we will be able to see this through.

@mado89, @s0nata, and @feliperodri, can you provide any update on when you will be able to complete your review? For some reason I thought we had reviews from at least one of you (@mado89) but we have none. I really don't want to keep @arcuri82 waiting much longer.

**Comment by arcuri82** [comment]
@gkthiruvathukal  thanks

Apart from the message of 29th April, it seems there has been no other communication from @feliperodri 

@mado89  has been a reviewer since January, but unfortunately has been postponing, month after month

I guess that, besides @s0nata, there is a need for a second reviewer.
From the list of the 26th of April, other possibilities could be:

* @UTH-Tuan
* @shanamatthews

**Comment by gkthiruvathukal** [comment]
@arcuri82, my pleasure. We will get there.

Apart from all of this, I also had to shut down for much of COVID-19 until the end of my academic year (May). So thank you for your patience.

Since I am not even getting a response from the two you  mentioned, I think I'm going to proceed with adding a reviewer with the hope that @s0nata will eventually  be able to complete a review for us.

**Comment by gkthiruvathukal** [comment]
@UTH-Tuan and @shanamatthews, would either or both of you be willing to contribute a review for this JOSS submission?

**Comment by gkthiruvathukal** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Re-invite a reviewer (if they can't update checklists)
@whedon re-invite @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to do a  dry run of accepting the paper and depositing with Crossref
@whedon accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

EiC TASKS

# Invite an editor to edit a submission (sending them an email)
@whedon invite @editor as editor

# Reject a paper
@whedon reject

# Withdraw a paper
@whedon withdraw

# Ask Whedon to actually accept the paper and deposit with Crossref
@whedon accept deposit=true

```

**Comment by UTH-Tuan** [comment]
> @UTH-Tuan and @shanamatthews, would either or both of you be willing to contribute a review for this JOSS submission?

@gkthiruvathukal I guess I can step in light of the situation. I'll admit that this is a new way of doing reviews for manuscripts so I may need some direction and guidance.

**Comment by gkthiruvathukal** [comment]
@UTH-Tuan I will be able to answer any questions about the review process.  Just so you know, we have reviewer guidelines that take you through the process at https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. I'm going to add you as a reviewer now!

**Comment by gkthiruvathukal** [comment]
@whedon add @UTH-Tuan as reviewer

**Comment by whedon** [comment]
OK, @UTH-Tuan is now a reviewer

**Comment by mado89** [comment]
> @arcuri82 So sorry about this. I am still hopeful we will be able to see this through.
> 
> @mado89, @s0nata, and @feliperodri, can you provide any update on when you will be able to complete your review? For some reason I thought we had reviews from at least one of you (@mado89) but we have none. I really don't want to keep @arcuri82 waiting much longer.

I am sorry for not posting my review more explicitily: 
In my opinion this tool needs some more work. Although investing quite some time into this review. I am still unable to perform a thorough review of this tool for the following reasons:

* Lack of gradle support in documentation (I am aware that most people use Maven. Maybe my issues are related to some other aspects)
* Lack of user documentation: For "white box" testing there is a document explaining a lot of technical details. However, in my opinion it lacks a "Step-by-step" like guide helping novice users of the tool performing their first test. Hence, I was unable to perform the test (also I just received a cryptic failure message, which I reported back to the author. There the review stopped). 
* Lack of OAuth support: I had a brief discussion with the author about this. However, we are in a disagreement about the importance of OAuth. If mentioned somewhere more prominent that OAuth is not supported than this could be ok for me. I discovered this draw back only after investing quite some time into the review and trying to set up a test for an application.

Unfortunately, this comments are somewhat hard to add to the standard review form. 
When these issues are somewhat addressed I am willing to again spend some time on this review (I spent several days trying to get the tool running).

**Comment by arcuri82** [comment]
I am here going to reply to the 3 points raised by @mado89 in the previous post

1. In the 6 months that @mado89 has been a reviewer, 4 issues ([https://github.com/EMResearch/EvoMaster/issues](https://github.com/EMResearch/EvoMaster/issues)) were open: 169, 170, 177 and 178. Unless I missed it, nothing about Gradle was ever mentioned.  Adding an example of Gradle import of a library is pretty trivial, and I ll do it after writing this response.

2. I guess this is referring to  [177](https://github.com/EMResearch/EvoMaster/issues/177), where @mado89 wrote `The instructions are much clearer now`. If there are parts that are still not clear, this does not seem reported in [177](https://github.com/EMResearch/EvoMaster/issues/177) (besides OAuth, discussed next).
Furthermore, regarding `(also I just received a cryptic failure message, which I reported back to the author. There the review stopped)`: I assume @mado89 is referring to [178](https://github.com/EMResearch/EvoMaster/issues/178). If that is the case, I am still waiting for my questions of 21st of May. It is impossible to debug such issue if no more info is provided. 

 3. On this point, I am in disagreement with @mado89. As such communication was done via email, I will post here my answer:

> On this, I do disagree: 
> - of all open-source REST API that I found (eg, then ones in https://github.com/EMResearch/EMB), none use OAuth
> - none of the industrial partners I collaborate with use OAuth (and they have SpringBoot REST APIs)
> - on API directories (eg https://www.programmableweb.com/  and https://www.programmableweb.com/) there are literally hundreds (if not thousands) of APIs not using OAuth. If you want to look at some numbers, check > https://www.programmableweb.com/news/spotting-api-security-trends-programmablewebs-api-directory/research/2018/01/02
> - no other test generation tool support OAuth
>
> We are going to support OAuth in the (hopefully not far) future, but it has not be a priority so far, as it is not so popular yet. There are other more important features to add first

**Comment by arcuri82** [comment]
Gradle example has now been added to the [documentation](https://github.com/EMResearch/EvoMaster/blob/master/docs/write_driver.md).

**Comment by arcuri82** [comment]
To clarify regarding OAuth: `If mentioned somewhere more prominent that OAuth is not supported than this could be ok for me`. As I wrote in [177](https://github.com/EMResearch/EvoMaster/issues/177), the fact that OAuth is not supported has been added to the documentation, in the section regarding Security. As I did not get any reply in 177, I thought it was OK. But I can add it now to the main readme.

**Comment by mado89** [comment]
@arcuri82 Perfect! I will try to have a look as soon as possible. Please ping me if you don't have a reply within 10 days

**Comment by UTH-Tuan** [comment]
@arcuri82 I have taken a look at the paper to understand the scope of the problem that the software addresses. I don't see any discussion relating to what others have done and how this tool differentiates itself.

**Comment by arcuri82** [comment]
@UTH-Tuan, from the [paper](https://github.com/openjournals/joss-papers/blob/joss.02153/joss.02153/10.21105.joss.02153.pdf):

> Related Work
>In the recent years, different techniques have been proposed for black-box testing of RESTful APIs (e.g., (Atlidakis, Godefroid, & Polishchuk, 2019)(Karlsson, Causevic, & Sundmark, 2020) (Viglianisi, Dallago, & Ceccato, 2020)(Ed-douibi, Cánovas Izquierdo, & Cabot, 2018)).
>However, at the time of this writing, EvoMaster appears to be the only one can do both black-box and white-box testing, and that is also released as open-source

Are you saying that you want to see that (very brief) section extended?

**Comment by UTH-Tuan** [comment]
@arcuri82 yes a brief extended version for those unfamiliar with the domain.

**Comment by arcuri82** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02153/joss.02153/10.21105.joss.02153.pdf)

**Comment by arcuri82** [comment]
@UTH-Tuan thanks. I have added some more info in the Related Work section. Right now the *paper.md* file has 974 words, where JOSS has the limit of [250-1000 words](https://joss.readthedocs.io/en/latest/submitting.html).

**Comment by UTH-Tuan** [comment]
@arcuri82 Thanks. I am also encountering issues in running the example. Either that or is it not clear how to run the example (based off what was presented in the Example doc). I get the error "Cannot communicate with remote EvoMaster Driver for the system under test at localhost:40100 Make sure the EvoMaster Driver for the system under test is running correctly". Please clarify and/or make the example more straightforward.

**Comment by arcuri82** [comment]
@UTH-Tuan in the main readme of the documentation: *A short video (5 minutes) shows the use of EvoMaster on one of the case studies in EMB.* [Link](https://www.youtube.com/watch?v=3mYxjgnhLEo&feature=youtu.be).
This should help better see how things are connected, besides the description in the [documentation](https://github.com/EMResearch/EvoMaster/blob/master/docs/whitebox.md).
 
Btw, could be better to create new GitHub issues in the [EvoMaster repository](https://github.com/EMResearch/EvoMaster/issues), instead of having the review here (see [JOSS documentation](https://joss.readthedocs.io/en/latest/reviewer_guidelines.html)).

**Comment by gkthiruvathukal** [comment]
@arcuri82 is right. The issue tracker is the best way to report issues on a submission, especially when it's on GitHub!  Many reviewers will post the link to the issue in the review as well, so it is easy to click through.

Thanks to @UTH-Tuan for helping with this review! We really appreciate it.

**Comment by gkthiruvathukal** [comment]
@arcuri82  Based on the open issue raised by @mado89, I assume you are still working to bring it to closure.

@mado89 and @UTH-Tuan,  please respond when you are ready to move forward with this submission. I will be nudging everyone every week or so.

**Comment by mado89** [comment]
Yes we are trying to resolve some issues I have with the software

**Comment by s0nata** [comment]
Dear everybody, thank you for your patience with me. Got significantly affected at work due to covid situation, and only recently was able to return to a schedule that premits time for the review.

For my review the only missing part is the evaluation, and @arcuri82 , by any chance is there and example project I can set up and run EvoMaster on? I saw the links in the docs for black- and white-box testing setups, they do describe some possible use cases, but I would find it extremely helpful if there was a follow-up script (as in theater play script) for "Hey, to see how our tool works on a sample project, build it, then use this small toy project to evaluate it on, you should get following results for scenarios A, B, and C". Having a docker image with a demo would be probably too resource-consuming to produce, so not asking for it.

**Comment by arcuri82** [comment]
Hi @s0nata,

   having a demo project would indeed be useful. Although these weeks I am quite busy with the starting of teaching semester, so might take a while before I have time to prepare it.

Anyway, for the current time, you can look at the following: from the [main readme](https://github.com/EMResearch/EvoMaster), there is a link to the [EMB](https://github.com/EMResearch/EMB) repository, which contains a collection of REST APIs with also drivers for white-box testing. From the main readme, there is also a link to a [YouTube video](https://www.youtube.com/watch?v=3mYxjgnhLEo&feature=youtu.be) showing how to run EM on one of such projects.

**Comment by gkthiruvathukal** [comment]
@arcuri82, Given that I, too, am working intensely to be prepared for classes this week (with lots of video recording to do, being online) I am going to nudge you & the reviewers in a week or two. As you know, I want to bring this submission to closure, especially  given how long you had to wait. 

@s0nata and @UTH-Tuan, please do let me know if @arcuri82's suggestion helps you to move forward with ticking the remaining boxes.

**Comment by arcuri82** [comment]
In the end, it took less than I thought.
I created a small example of API, available at [rest-api-example](https://github.com/EMResearch/rest-api-example).
I also made a [video](https://youtu.be/ORxZoYw7LnM) showing how to implement an`EMDriver` to enable white-box testing with EvoMaster.
Video is now linked from the main Readme documentation.

**Comment by mado89** [comment]
The video looks good. I took some notes while watchin it and will share them with you. Maybe they could be helpful for potential users of EvoMaster. 
Using the video I think I will be able to do my own testing using Evo Master.

Just one thing: Right at the end of the video either the audio cuts off or you stop speaking. Either way it is a bit hard to follow what you are doing (but I think I might be able to reproduce it using your other video about blackbox testing).

**Comment by kthyng** [comment]
Hi all! As the rotating associate editor in chief this week, I just wanted to check in on this issue. I see unchecked review boxes above. How is your review going @mado89, @feliperodri, @s0nata, @UTH-Tuan? Thanks all for your work!!

**Comment by gkthiruvathukal** [comment]
@kthyng I think we're really close on this one. I also notice there is no checklist for @UTH-Tuan, who was added as a reviewer. I don't think @feliperodri is still available for this review, which is one of the reasons we added @UTH-Tuan. So I think we only need @mado89 , @s0nata , and @UTH-Tuan to indicate whether they are waiting on any further changes.

**Comment by kthyng** [comment]
@gkthiruvathukal ah! I see. Ok, I will remove the one reviewer and edit the checklist to be for @UTH-Tuan instead. When reviewer changes happen after starting the review issue, some of this needs to happen by hand. Please let me know if anything remains incorrect after I'm done.

**Comment by whedon** [comment]
OK, @feliperodri is no longer a reviewer

**Comment by UTH-Tuan** [comment]
@gkthiruvathukal @kthyng how do I generate a checklist? I am sorry but this my first time doing this type of review without any direction  (I am more accustomed to the traditional paper review).

**Comment by kthyng** [comment]
@UTH-Tuan You now have a checklist above to use. So the review is checklist-driven, as above, but as issues come up, please open issues on the software repo to describe and discuss, and link to this review issue so that we can track related issues. Let me or @gkthiruvathukal know if you have any questions. Thanks!

**Comment by UTH-Tuan** [comment]
@kthyng @gkthiruvathukal I see it now. Earlier I didn't have a checklist. I initially thought I had generate it . Thanks!

**Comment by UTH-Tuan** [comment]
@kthyng @gkthiruvathukal how do i check off the items from the checklist?

**Comment by kthyng** [comment]
@UTH-Tuan If you cannot, you probably need to accept your reviewer invitation. I'll reinvite you.

**Comment by kthyng** [comment]
@whedon re-invite @UTH-Tuan  as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@uth-tuan  please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by UTH-Tuan** [comment]
@kthyng I get the following message "Sorry, we couldn't find that repository invitation. It is possible that the invitation was revoked or that you are not logged into the invited account."

**Comment by kthyng** [comment]
@whedon remove @UTH-Tuan as reviewer

**Comment by whedon** [comment]
OK, @UTH-Tuan is no longer a reviewer

**Comment by kthyng** [comment]
@whedon invite @UTH-Tuan as reviewer

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by kthyng** [comment]
@whedon add @UTH-Tuan as reviewer

**Comment by whedon** [comment]
OK, @UTH-Tuan is now a reviewer

**Comment by kthyng** [comment]
@UTH-Tuan please try again and see if it works for you

**Comment by arfon** [comment]
@whedon re-invite @UTH-Tuan as reviewer

**Comment by whedon** [comment]
@uth-tuan already has access.

**Comment by mado89** [comment]
Just wanted to ask if there is still work left from my side. I've ticked of all my boxes some time ago. While I think there is room for improvement concerning usability I think this tool can be accepted for publication in JOSS.

**Comment by gkthiruvathukal** [comment]
@mado89 I am inclined to move toward acceptance, unless there are any additional concerns from @UTH-Tuan.

**Comment by UTH-Tuan** [comment]
I would move toward acceptance. no concerns

**Comment by gkthiruvathukal** [comment]
@UTH-Tuan Thanks, this will help me to move toward the next steps. I'll be following up shortly.

**Comment by gkthiruvathukal** [comment]
Thanks, @mado89 and @UTH-Tuan. First, my apologies for to @arcuri82 for the delay as the past few weeks have been pretty intense leading to the end of semester. Thanks for your incredible patience during this process.

@arcuri82 Please do the following:

- [ ] Make a tagged release of your software, and list the version tag of the archived version here. 
- [ ] Archive the reviewed software in Zenodo 
- [ ] Check the Zenodo deposit has the correct metadata, this includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it); you may also add the authors' ORCID.
- [ ] List the Zenodo DOI of the archived version here.

**Comment by arcuri82** [comment]
@gkthiruvathukal: thanks. I will do it in the next coming days

**Comment by arcuri82** [comment]
Hi,
* I made a new release [1.1.0](https://github.com/EMResearch/EvoMaster/releases/tag/v1.1.0)
* I created a new entry on Zenodo, [4300745](https://zenodo.org/record/4300745). However, that is the first time I use Zenodo, so not sure if I did it correctly

**Comment by arcuri82** [comment]
@gkthiruvathukal: is there any news on this? On Thursday the 7th, it is going to be 1 year since this article was first submitted :(

**Comment by gkthiruvathukal** [comment]
Hi @arcuri82. Yes, we're definitely heading to acceptance. I hoped to process it before taking a much-needed winter break. I'll finalize it today or tomorrow. Thanks for your patience! (As you know, we lost some reviewers during the review process but everything is done now. It does not normally take this long!)

**Comment by gkthiruvathukal** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Re-invite a reviewer (if they can't update checklists)
@whedon re-invite @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# All commands can be run on a non-default branch, to do this pass a custom 
# branch name by following the command with `from branch custom-branch-name`.
# For example:

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to do a dry run of accepting the paper and depositing with Crossref
@whedon accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

EiC TASKS

# Invite an editor to edit a submission (sending them an email)
@whedon invite @editor as editor

# Reject a paper
@whedon reject

# Withdraw a paper
@whedon withdraw

# Ask Whedon to actually accept the paper and deposit with Crossref
@whedon accept deposit=true

```

**Comment by gkthiruvathukal** [comment]
@whedon set 10.5281/zenodo.4300745 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4300745" target="_blank">10.5281/zenodo.4300745</a> is the archive.

**Comment by gkthiruvathukal** [comment]
@whedon set 1.1.0 as release

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by gkthiruvathukal** [comment]
@whedon set 1.1.0 as version

**Comment by whedon** [comment]
OK. 1.1.0 is the version.

**Comment by gkthiruvathukal** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02153/joss.02153/10.21105.joss.02153.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02153/joss.02153/10.21105.joss.02153.pdf) :page_facing_up: :point_left:

**Comment by gkthiruvathukal** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/icst.2018.00046 is OK
- 10.1007/978-3-319-66299-2_1 is OK
- 10.1016/j.infsof.2018.05.003 is OK
- 10.1109/qrs.2017.11 is OK
- 10.1145/3293455 is OK
- 10.1145/3321707.3321815 is OK
- 10.1145/3321707.3321732 is OK
- 10.1109/ICST46399.2020.00025 is OK
- 10.1145/2379776.2379787 is OK
- 10.1109/ICSE.2019.00083 is OK
- 10.1109/ICST46399.2020.00023 is OK
- 10.1109/ICST46399.2020.00024 is OK
- 10.1109/EDOC.2018.00031 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2010

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2010, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kyleniemeyer** [comment]
Hi @arcuri82, I'm just doing some final checks on your submission before publishing.

I noticed some issues with in-text citations in your article (not using a single command when citing multiple things), and made a small PR to fix this: https://github.com/EMResearch/EvoMaster/pull/232

Can you merge this? I will then publish the article. There is no need to create a new Zenodo archive. Thanks!

**Comment by arcuri82** [comment]
@gkthiruvathukal: thanks
@kyleniemeyer: thanks for PR. it is merged now

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02153/joss.02153/10.21105.joss.02153.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02153/joss.02153/10.21105.joss.02153.pdf) :page_facing_up: :point_left:

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1345826731362562048) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2011
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02153](https://doi.org/10.21105/joss.02153)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by arcuri82** [comment]
thanks. generated PDF seems fine

**Comment by kyleniemeyer** [comment]
Congrats @arcuri82 on your article's publication in JOSS!

Many thanks to @mado89, @s0nata, and @UTH-Tuan for reviewing this, and @gkthiruvathukal for editing.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02153/status.svg)](https://doi.org/10.21105/joss.02153)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02153">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02153/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02153/status.svg
   :target: https://doi.org/10.21105/joss.02153
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02153/status.svg)](https://doi.org/10.21105/joss.02153)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
